feat: add ability to set schedule for sagemaker llm endpoints

Rob-Powell opened the pull request.

QuinnGT committed:
Add support for cohere embedding models

QuinnGT committed:
Move static values to Enum. Create new provider functions.

BrianFarnhill commented on cli/aws-cron-validator.ts:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    private static minuteRegex(): string {
+        return `^(${this.commonRegex(this.minuteExp)})$`;
```
nit: Given that the const `minuteExp` isn't used anywhere other than here, what have you separated it out to be its own constant? I see some of them are re-used but the ones that aren't might be more readable if you did them inline.

BrianFarnhill commented on cli/aws-cron-validator.ts:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    // Build up regex's incrementally improving readability of the above expressions
+    private static rangeRegex(values: string): string {
```
nit: I would consider getting these to return `RegExp` objects to save you wrapping a lot of the constructor calls above when you get strings out here. For the cases where you need to re-use the string value as part of another regex below you can reference that like `this.rangeRegex.source` which is still pretty clean

BrianFarnhill commented on cli/aws-cron-validator.ts:
```diff
@@ -0,0 +1,109 @@
+
+export class AWSCronValidator {
+
+    private static minuteExp = `(0?[0-9]|[1-5][0-9])`; // [0]0-59
```
nit: consider moving these constants to an external file to improve readability in here.

BrianFarnhill commented on cli/aws-cron-validator.ts:
```diff
@@ -0,0 +1,109 @@
+    private static yearRegex(): string {
+        return `^(${this.commonRegex(this.yearExp)})$`;
+    }
+}
```
nit: leave an emtpy line at the end of your files - that way if you add more to it you don't get a line change on the last line, even though the content of that line won't look any different (thats why the little red icon thing appears here on github)

BrianFarnhill commented on cli/magic-config.ts:
```diff
@@ -13,6 +13,8 @@ import {
 import { LIB_VERSION } from "./version.js";
 import * as fs from "fs";
+import { AWSCronValidator } from "./aws-cron-validator"
+import { AWSCronValidatorSimple } from "./aws-cron-validator-simple"
```
I assume this line can be killed off since as I don't see a reference to that class anywhere?

BrianFarnhill commented on cli/magic-config.ts:
```diff
@@ -68,6 +70,9 @@ const embeddingModels = [
       options.enableSagemakerModels = config.llms?.sagemaker
         ? config.llms?.sagemaker.length > 0
         : false;
+      options.enableSagemakerModelsSchedule = config.llms?.sagemakerSchedule?.enabled;
```
Can you test for the presence of `config.llms?.sagemakerSchedule` first, and then check the properties within that? I'm thinking of a scenario here I say set start and end times but don't set enabled = true ... or is that a scenario this should support? If so ignore this.

BrianFarnhill commented on cli/magic-config.ts:
```diff
@@ -197,6 +202,67 @@ async function processCreateOptions(options: any): Promise<void> {
+          return true
+        }
+        catch (error) {
+          if (error instanceof Error){
```
Would you trap and handle the `AWSCronError` separately here? If not, does that error class need to exist?

BrianFarnhill commented on cli/magic-config.ts:
```diff
@@ -197,6 +202,67 @@ async function processCreateOptions(options: any): Promise<void> {
+          if (error instanceof Error){
+            return error.message
+          }
+          console.log("hmmm")
```
hmmmmm indeed :)

BrianFarnhill commented on lib/models/index.ts:
```diff
@@ -32,6 +33,19 @@ export class Models extends Construct {
+      });
+      
+      schedulerRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSageMakerFullAccess'));
+      this.suppressCdkNagWarningForEndpointRole(schedulerRole);
```
`IAM4` and `IAM5` are typically pretty broad ones to disable without a specific `appliesTo` clause (if someone else grabs the role elsewhere in a stack and adds or changes it, the rule is already modified). I would consider being more careful with that suppression.

BrianFarnhill commented on lib/models/index.ts:
```diff
@@ -63,7 +77,13 @@ export class Models extends Construct {
       });
-
+  
+      if (
```
Rather than dumping this if statement in to each model, would it be possible to iterate over the collection stored in the `models` variable at the end? Something like:

```
if (models.length > 0 && props.config.llms?.sagemakerSchedule?.enabled && schedulerRole !== undefined) {
    models.forEach((model) => {
        createStartSchedule(this, id, model.cfnEndpoint, schedulerRole, props.config.llms?.sagemakerSchedule?.sagemakerStartSchedule);
        createStopSchedule(this, id, modelcfnEndpoint, schedulerRole, props.config.llms?.sagemakerSchedule?.sagemakerStopSchedule);
    });
}
```

That way you can do it once at the end and catch all of the models?

BrianFarnhill commented on lib/models/sagemaker-schedule.ts:
```diff
@@ -0,0 +1,48 @@
+import * as sagemaker from "aws-cdk-lib/aws-sagemaker";
+import { Construct } from 'constructs';
+
+export function createStartSchedule(scope: Construct, id: string, sagemakerEndpoint: sagemaker.CfnEndpoint, role: iam.Role, scheduleExpression: string) {
```
To follow suit with how other CDK constructs do it, rather than put all the params here as positional args, I would build a "Props" type and use that:

```
export interface ScheduleProps {
    sagemakerEndpoint: sagemaker.CfnEndpoint;
    role: iam.Role;
    scheduleExpression: string;
}

export function createStartSchedule(scope: Construct, id: string, props: ScheduleProps) {
```

BrianFarnhill commented on lib/models/sagemaker-schedule.ts:
```diff
@@ -0,0 +1,48 @@
+
+export function createStartSchedule(scope: Construct, id: string, sagemakerEndpoint: sagemaker.CfnEndpoint, role: iam.Role, scheduleExpression: string) {
+  const startSchedule = new scheduler.CfnSchedule(scope, `${id}-${sagemakerEndpoint.endpointName}-startScheduler`, {
+    name: `StartScheduler-${id}-${sagemakerEndpoint.endpointName}`,
```
With the ID in here, do you need that in the string? Is there a scenario where you end up with multiple start schedules for the same endpoint name?

BrianFarnhill commented on lib/models/sagemaker-schedule.ts:
```diff
@@ -0,0 +1,48 @@
+
+export function createStopSchedule(scope: Construct, id: string, sagemakerEndpoint: sagemaker.CfnEndpoint, role: iam.Role, scheduleExpression: string) {
+  const stopSchedule = new scheduler.CfnSchedule(scope, `${id}-${sagemakerEndpoint.endpointName}-stopScheduler`, {
+    name: `StopScheduler-${id}-${sagemakerEndpoint.endpointName}`,
```
Same question as above about the need to put ID in this string

BrianFarnhill commented on lib/models/sagemaker-schedule.ts:
```diff
@@ -0,0 +1,48 @@
+  return startSchedule;
+}
+
+export function createStopSchedule(scope: Construct, id: string, sagemakerEndpoint: sagemaker.CfnEndpoint, role: iam.Role, scheduleExpression: string) {
```
Same comment on the props usage here as above - further to that though, could this just be a `createSchedule` call that takes a property like `scheduleType: 'START' | 'STOP'` that you can use to toggle on the differences? I only see the description and the target ARN that differ - could save you duplicating some of the rest of the logic.

BrianFarnhill commented on lib/shared/types.ts:
```diff
@@ -84,6 +84,11 @@ export interface SystemConfig {
   llms: {
     sagemaker: SupportedSageMakerModels[];
+    sagemakerSchedule?: {
+      enabled?: boolean;
```
I know i commented on it way at the start, but again here - is there a scenario where enabled is ommited but you would still pass the schedules?

BrianFarnhill reviewed (commented):
Some super nit picky stuff in here so take/leave any of it as you choose. From a functional stand point it looks great, there's a few improvements I would consider for readability and consistency but nothing dramatic. Nice code mate.

QuinnGT committed:
Add Enum to query, chunks and workspaces

QuinnGT committed:
Fix error with API calls to provider and model

QuinnGT committed:
Add Titan Multimodal Embeddings-No image support

QuinnGT committed:
Fix default embedding model prompt on create

QuinnGT committed:
Removed mistake on hardcoded enum values

QuinnGT committed:
Add condition for embedding to only show on aurora or opensearch

QuinnGT committed:
Update task with enum

Commit:
convert webcrawler to fargate container

Commit:
clean-up of unused code

Commit:
remove dead code

massi-ang committed:
feat(message_api): moving to message API for Claude models

massi-ang committed:
feat(message_api): only yield on relevant chunks

massi-ang committed:
fix(chat): fixing chat for streamed content

massi-ang committed:
fix(chat): only send messages when token length > 0

massi-ang committed:
fix(cleanup): remove unused code

massi-ang committed:
fix(streaming): fixing issues with streamed responses

massi-ang committed:
fix(ui): replaced react-icons with cloudscape icons

massi-ang committed:
feat(claude_mm): adding multimodal chat for Claude

redesign of the Idefics Model Interface to accept multiple model via adapters.

massi-ang committed:
feat(multimodal): multiple images upload

massi-ang committed:
feat(multimodal): fixes to initialize the adapters

massi-ang committed:
feat(mistral): support for mistral and mixtral models

fixes to the LLama2 and mistral prompts

massi-ang committed:
feat(mistral): adapter registration regex for mistral

massi-ang committed:
feat(mistral): separate adapter

massi-ang committed:
fix(mistral): update of registry location

surukonda committed:
enhance cli to capture existing vpc details

dairiley committed:
fix: remove length > 0 condition for idefics path suppressions so they are always suppressed

massi-ang committed:
Merge pull request #403 from dairiley/fix_nag_suppression_idefics

fix: remove length > 0 condition for idefics path suppressions

Commit:
consider question skipping status while validating vpc id

massi-ang committed:
fix(claude multimodal): key error

massi-ang committed:
fix: sorting of SQS messages int the batch

massi-ang committed:
fix(mistral): correct parameters

massi-ang committed:
fix(streaming): better logic

massi-ang committed:
fix(claude image): continue to use multimodal mode in a chat with images

massi-ang committed:
fix(cw role): setting for automatic creation of cloudwatch role for apigw

massi-ang committed:
fix(multimodal): take into account the capabilties of the selected model

When deciding which interface to use (multimodal or langchain), consider also the InputModalities
of the selected model

massi-ang committed:
fix(claude prompt): improved claude prompts and fix to model interface selection

Rob-Powell committed:
feat: adding schedule option for sagemaker models
